beater/api/mux.go

```diff
@@ -179,17 +180,29 @@ func (r *routeBuilder) rumAgentConfigHandler() (request.Handler, error) {
 type middlewareFunc func(*config.Config, *authorization.Handler, map[request.ResultID]*monitoring.Int) []middleware.Middleware
 
 func agentConfigHandler(cfg *config.Config, authHandler *authorization.Handler, middlewareFunc middlewareFunc) (request.Handler, error) {
-	var client kibana.Client
-	if cfg.Kibana.Enabled {
-		client = kibana.NewConnectingClient(&cfg.Kibana)
+	var h request.Handler
+	mw := middlewareFunc(cfg, authHandler, agent.MonitoringMap)
+	if cfg.AgentConfigs != nil {
+		// Direct agent configuration is present, disable communication
+		// with kibana.
+		f := agentcfg.NewDirectFetcher(cfg.AgentConfigs)
+		h = agent.NewHandler(f, cfg.KibanaAgentConfig, cfg.DefaultServiceEnvironment)
+	} else {
+		var client kibana.Client
+		if cfg.Kibana.Enabled {
+			client = kibana.NewConnectingClient(&cfg.Kibana)
+		}
+
+		f := agentcfg.NewFetcher(client, cfg.KibanaAgentConfig.Cache.Expiration)
```

Comment:
When we come to fixing up the jaeger bits, we'll probably want to extract this logic into a function in the `beater` package. Then we can pass the fetcher into the places it's needed: `beater/mux`, `beater/jaeger`, and the gRPC server in the `beater` package itself.

Reply:
Ended up putting it in `agentcfg`. The logic was entirely based on stuff in the `config` package, and it was returning a `Fetcher`. Happy to put it elsewhere of course.